src/blockedtuple.jl

```diff
@@ -0,0 +1,107 @@
+
+# Base interface
+Base.axes(bt::AbstractBlockTuple) = (blockedrange([blocklengths(bt)...]),)
+
+Base.deepcopy(bt::AbstractBlockTuple) = deepcopy.(bt)
+
+Base.firstindex(::AbstractBlockTuple) = 1
+
+Base.getindex(bt::AbstractBlockTuple, i::Integer) = Tuple(bt)[i]
+Base.getindex(bt::AbstractBlockTuple, r::AbstractUnitRange) = Tuple(bt)[r]
+Base.getindex(bt::AbstractBlockTuple, b::Block{1}) = blocks(bt)[Int(b)]
+function Base.getindex(bt::AbstractBlockTuple, br::BlockRange{1})
+  r = Int.(br)
+  T = unspecify_type_parameters(typeof(bt))
+  flat = Tuple(bt)[blockfirsts(bt)[first(r)]:blocklasts(bt)[last(r)]]
+  return T{blocklengths(bt)[r]}(flat)
+end
+function Base.getindex(bt::AbstractBlockTuple, bi::BlockIndexRange{1})
+  return bt[Block(bi)][only(bi.indices)]
+end
+
+Base.iterate(bt::AbstractBlockTuple) = iterate(Tuple(bt))
+Base.iterate(bt::AbstractBlockTuple, i::Int) = iterate(Tuple(bt), i)
+
+Base.length(bt::AbstractBlockTuple) = length(Tuple(bt))
+
+Base.lastindex(bt::AbstractBlockTuple) = length(bt)
+
+function Base.map(f, bt::AbstractBlockTuple)
+  return unspecify_type_parameters(typeof(bt)){blocklengths(bt)}(map(f, Tuple(bt)))
+end
+
+# Broadcast interface
+Base.broadcastable(bt::AbstractBlockTuple) = bt
+struct AbstractBlockTupleBroadcastStyle{BlockLengths,BT} <: Broadcast.BroadcastStyle end
+function Base.BroadcastStyle(T::Type{<:AbstractBlockTuple})
+  return AbstractBlockTupleBroadcastStyle{blocklengths(T),unspecify_type_parameters(T)}()
+end
+
+# BroadcastStyle is not called for two identical styles
+function Base.BroadcastStyle(
+  ::AbstractBlockTupleBroadcastStyle, ::AbstractBlockTupleBroadcastStyle
+)
+  throw(DimensionMismatch("Incompatible blocks"))
+end
+function Base.copy(
+  bc::Broadcast.Broadcasted{AbstractBlockTupleBroadcastStyle{BlockLengths,BT}}
+) where {BlockLengths,BT}
+  return BT{BlockLengths}(bc.f.((Tuple.(bc.args))...))
+end
+
+# BlockArrays interface
+function BlockArrays.blockfirsts(bt::AbstractBlockTuple)
+  return (0, cumsum(Base.front(blocklengths(bt)))...) .+ 1
+end
+
+function BlockArrays.blocklasts(bt::AbstractBlockTuple)
+  return cumsum(blocklengths(bt)[begin:end])
+end
+
+BlockArrays.blocklength(bt::AbstractBlockTuple) = length(blocklengths(bt))
+
+BlockArrays.blocklengths(bt::AbstractBlockTuple) = blocklengths(typeof(bt))
+
+function BlockArrays.blocks(bt::AbstractBlockTuple)
+  bf = blockfirsts(bt)
+  bl = blocklasts(bt)
+  return ntuple(i -> Tuple(bt)[bf[i]:bl[i]], blocklength(bt))
+end
+
+#
+# =====================================  BlockedTuple  =====================================
+#
+struct BlockedTuple{BlockLengths,Flat} <: AbstractBlockTuple
+  flat::Flat
+
+  function BlockedTuple{BlockLengths}(flat::Tuple) where {BlockLengths}
+    length(flat) != sum(BlockLengths) && throw(DimensionMismatch("Invalid total length"))
+    return new{BlockLengths,typeof(flat)}(flat)
+  end
+end
+
+# TensorAlgebra Interface
+tuplemortar(tt::Vararg{Tuple}) = BlockedTuple{length.(tt)}(flatten_tuples(tt))
```

Comment:
```suggestion
tuplemortar(tt::Tuple{Vararg{Tuple}}) = BlockedTuple{length.(tt)}(flatten_tuples(tt))
```
`mortar` expects an array of arrays:
```julia
julia> using BlockArrays

julia> mortar([[1, 2], [3, 4]])
2-blocked 4-element BlockVector{Int64}:
 1
 2
 ─
 3
 4

julia> mortar([1, 2], [3, 4])
ERROR: MethodError: no method matching _BlockArray(::Vector{Int64}, ::Vector{Int64})
[...]
```
so I think we should match that convention and require users pass a Tuple of Tuples.

Reply:
It requires more typing, but I think the reasoning behind that design is that `mortar` can be thought of as the inverse of `blocks`, i.e.:
```julia
julia> v = BlockVector(1:4, [2, 2])
2-blocked 4-element BlockVector{Int64, Vector{UnitRange{Int64}}, Tuple{BlockedOneTo{Int64, Vector{Int64}}}}:
 1
 2
 ─
 3
 4

julia> mortar(blocks(v))
2-blocked 4-element BlockVector{Int64, Vector{UnitRange{Int64}}, Tuple{BlockedOneTo{Int64, Vector{Int64}}}}:
 1
 2
 ─
 3
 4
```